File: src/main/java/org/panteleyev/money/app/AboutDialog.java
/*
 Copyright (C) 2017, 2018, 2019, 2020, 2021, 2022 Petr Panteleyev

 This program is free software: you can redistribute it and/or modify it under the
 terms of the GNU General Public License as published by the Free Software
 Foundation, either version 3 of the License, or (at your option) any later version.

 This program is distributed in the hope that it will be useful, but WITHOUT ANY
 WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

 You should have received a copy of the GNU General Public License along with this
 program. If not, see <https://www.gnu.org/licenses/>.
 */
package org.panteleyev.money.app;

import javafx.scene.control.ButtonType;
import javafx.scene.image.ImageView;
import org.panteleyev.fx.BaseDialog;
import org.panteleyev.fx.Controller;

import java.time.LocalDate;
import java.util.ResourceBundle;

import static org.panteleyev.fx.BoxFactory.vBox;
import static org.panteleyev.fx.FxUtils.fxString;
import static org.panteleyev.fx.LabelFactory.label;
import static org.panteleyev.money.app.GlobalContext.settings;
import static org.panteleyev.money.app.Styles.BIG_SPACING;
import static org.panteleyev.money.app.Styles.SMALL_SPACING;
import static org.panteleyev.money.app.Styles.STYLE_ABOUT_LABEL;

final class AboutDialog extends BaseDialog<Object> {

    private record BuildInformation(String version, String timestamp) {
        static BuildInformation load() {
            var bundle = ResourceBundle.getBundle("org.panteleyev.money.buildInfo");
            return new BuildInformation(
                    bundle.getString("version"),
                    bundle.getString("timestamp")
            );
        }
    }

    static final String APP_TITLE = "Money Manager";

    private static final String RUNTIME = System.getProperty("java.vm.version") + " " + System.getProperty("os.arch");
    private static final String VM = System.getProperty("java.vm.name") + " by " + System.getProperty("java.vm.vendor");
    private static final BuildInformation BUILD = BuildInformation.load();

    AboutDialog(Controller owner) {
        super(owner, settings().getAboutDialogCssFileUrl());

        setHeaderText(APP_TITLE);
        setGraphic(new ImageView(Images.APP_ICON));

        setTitle("About " + APP_TITLE);

        var aboutLabel = label(fxString(APP_TITLE, " ") + BUILD.version());
        aboutLabel.getStyleClass().add(STYLE_ABOUT_LABEL);

        var vBox = vBox(BIG_SPACING,
                vBox(SMALL_SPACING,
                        aboutLabel,
                        label("Built on " + BUILD.timestamp())
                ),
                vBox(SMALL_SPACING,
                        label("Runtime version: " + RUNTIME),
                        label("VM: " + VM)
                ),
                vBox(SMALL_SPACING,
                        label("Copyright (C) 2016, " + LocalDate.now().getYear() + ", Petr Panteleyev")
                )
        );

        getDialogPane().setContent(vBox);
        getDialogPane().getButtonTypes().addAll(ButtonType.OK);
    }
}

File: src/main/kotlin/org/panteleyev/money/AboutDialog.kt
/*
 * Copyright (c) 2017, Petr Panteleyev <petr@panteleyev.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

package org.panteleyev.money

import javafx.geometry.Insets
import javafx.scene.control.ButtonType
import javafx.scene.control.Label
import javafx.scene.image.ImageView
import javafx.scene.layout.BorderPane
import javafx.scene.layout.GridPane
import javafx.scene.layout.VBox
import org.panteleyev.utilities.fx.BaseDialog
import java.util.ResourceBundle

class AboutDialog : BaseDialog<Any>(MainWindowController.DIALOGS_CSS) {
    companion object {
        const val APP_TITLE = "Money Manager"
        const val COPYRIGHT = "Copyright (c) 2016, 2017, Petr Panteleyev"
        private val BUILD_INFO = "org.panteleyev.money.res.buildInfo"
    }

    init {
        title = APP_TITLE

        val icon = ImageView(Images.APP_ICON).apply {
            fitWidth = 48.0
            fitHeight = 48.0
        }

        val rb: ResourceBundle = ResourceBundle.getBundle(BUILD_INFO)

        val grid = GridPane().apply {
            styleClass.add(Styles.GRID_PANE)
            addRow(0, Label("Version:"), Label(rb.getString("version")))
            addRow(1, Label("Build:"), Label(rb.getString("timestamp")))
        }

        val vBox = VBox(10.0,
                Label(APP_TITLE).apply { styleClass.add(Styles.ABOUT_APP_TITLE_LABEL) },
                Label(COPYRIGHT).apply { styleClass.add(Styles.ABOUT_LABEL) },
                grid)

        val pane = BorderPane().apply {
            left = icon
            center = vBox
        }

        BorderPane.setMargin(vBox, Insets(0.0, 0.0, 0.0, 10.0))

        dialogPane.content = pane
        dialogPane.buttonTypes.add(ButtonType.OK)
    }
}
